src/webgpu/api/validation/encoding/cmds/compute_pass.spec.ts

```diff
@@ -0,0 +1,135 @@
+    const descriptor: GPUBufferDescriptor = {
+      size: data.byteLength,
+      usage: GPUBufferUsage.INDIRECT | GPUBufferUsage.COPY_DST,
+    };
+
+    if (state === 'invalid') {
+      descriptor.usage = 0xffff; // Invalid GPUBufferUsage
+    }
+
+    this.device.pushErrorScope('validation');
+    const buffer = this.device.createBuffer(descriptor);
+    this.device.popErrorScope();
+
+    if (state === 'valid') {
+      this.queue.writeBuffer(buffer, 0, data);
+    }
+
+    if (state === 'destroyed') {
+      buffer.destroy();
+    }
+
+    return buffer;
+  }
+}
+
+export const g = makeTestGroup(F);
+
+g.test('set_pipeline')
+  .params(poptions('state', ['valid', 'invalid'] as const))
+  .fn(t => {
+    const pipeline = t.createComputePipeline(t.params.state);
+    const { encoder, finish } = t.createEncoder('compute pass');
+    encoder.setPipeline(pipeline);
+    t.expectValidationError(() => {
+      finish();
+    }, t.params.state === 'invalid');
+  });
+
+g.test('dispatch_sizes')
+  .params(
+    params()
+      .combine(poptions('dispatchType', ['direct', 'indirect'] as const))
+      .combine(
+        poptions('workSizes', [
+          [0, 0, 0],
+          [1, 1, 1],
+          // TODO: Add a test for a large workSize when the upper limit has been decided.
+        ] as [number, number, number][])
+      )
+  )
+  .fn(t => {
+    const pipeline = t.createNoOpComputePipeline();
+    const [x, y, z] = t.params.workSizes;
+    const { encoder, finish } = t.createEncoder('compute pass');
+    encoder.setPipeline(pipeline);
+    if (t.params.dispatchType === 'direct') {
+      encoder.dispatch(x, y, z);
+    } else if (t.params.dispatchType === 'indirect') {
+      encoder.dispatchIndirect(t.createIndirectBuffer('valid', new Uint32Array([x, y, z])), 0);
+    }
+    t.queue.submit([finish()]);
+  });
+
+const kBufferData = new Uint32Array(6).fill(1);
+g.test('indirect_dipatch_buffer')
+  .params(
+    params()
+      .combine(poptions('state', ['valid', 'invalid', 'destroyed'] as const))
+      .combine(
+        poptions('offset', [
+          0, // valid for 'valid' buffers
+          Uint32Array.BYTES_PER_ELEMENT, // valid for 'valid' buffers
+          kBufferData.byteLength - 3 * Uint32Array.BYTES_PER_ELEMENT, // valid for 'valid' buffers
+          1, // invalid, non-multiple of 4 offset
+          kBufferData.byteLength - Uint32Array.BYTES_PER_ELEMENT, // invalid, last element outside buffer
+        ])
+      )
+  )
+  .fn(t => {
+    const { state, offset } = t.params;
+    const pipeline = t.createNoOpComputePipeline();
+    const buffer = t.createIndirectBuffer(state, kBufferData);
+    const { encoder, finish } = t.createEncoder('compute pass');
+    encoder.setPipeline(pipeline);
+    t.expectValidationError(() => {
+      encoder.dispatchIndirect(buffer, offset);
+      t.queue.submit([finish()]);
+    }, state !== 'valid' || offset % 4 !== 0 || offset + 3 * Uint32Array.BYTES_PER_ELEMENT > kBufferData.byteLength);
+  });
```

Comment:
@kainino0x do we need to test that `dispatch` requires a valid pipelines with the correct bingroups or will it be in a different test?

Reply:
pipeline-bindgroup matching tests are going to be pretty involved and probably can be shared between compute and render, so probably a different file.